### PR TITLE
[CDEC-320] Investigate Cbor serialisation failure in CI. 

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

The CBOR serialisation tests in `Test.Pos.Cbor.CborSpec` were failing very occasionally running `binaryTests` on `Attributes X1`. One of the `binaryTests` is a canonicity test that uses `perturbCanonicity`. In this case it was duplicating the last element of the `Map` underlying `UnparsedFields`. Due to a small bug in `deserializeOrFail` that didn't return all leftover bytes from the input this was decoded fine, breaking canonicity. This has been fixed and tested against various previously failing seeds.

## Linked issue

[CDEC-320](https://iohk.myjetbrains.com/youtrack/issue/CDEC-320)

## Type of change

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [~] 🛠 New feature (non-breaking change which adds functionality)
- [~] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [~] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [~] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [~] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/src/Test/Pos/Helpers.hs [hunk 1]
```diff
@@ -56,22 +56,25 @@ cborFlatTermValid = property . validFlatTerm . toFlatTerm . encode
...
-    let sa = serialize a
-    sa' <- R.serialise <$> perturbCanonicity (R.deserialise sa)
-    let out = decodeFull @a $ sa'
-    pure $ case out of
-        -- perturbCanonicity may have not changed anything. Decoding can
-        -- succeed in this case.
-        Right a' -> counterexample (show a') (sa == sa')
-        -- It didn't decode. The error had better be a canonicity violation.
-        Left err -> counterexample (show err) (isCanonicityViolation err)
+  let sa = serialize a
```
Comment (erikd) [inline lib/src/Test/Pos/Helpers.hs]: At least part of the change  here seems to be whitespace, which makes seeing the actual functional changes a little difficult. Also, the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) for cardano-sl say "4 space indents". I use 2 space for my own projects, but here I go with 4.

Comment (erikd) [inline lib/src/Test/Pos/Helpers.hs]: At least part of the change  here seems to be whitespace, which makes seeing the actual functional changes a little difficult. Also, the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) for cardano-sl say "4 space indents". I use 2 space for my own projects, but here I go with 4.

**Comment by erikd** [comment]
This looks good to me, i've only read the Cbor code in passing. Specifically asking for a review from  @arybczak .

**Comment by arybczak** [comment]
> Due to a small bug in deserializeOrFail that didn't return all leftover bytes from the input this was decoded fine

Isn't that the problem that needs fixing?

Canonicity test, for a value `A`, is supposed to take its serialised form `SA` and perturb it randomly into `SA'` that deserialises into `A'` such that `A = A'`. If `A'` is not equal to `A`, the problem that needs to be fixed is elsewhere.

**Comment by arybczak** [comment]
> that deserialises into `A'` such that `A = A'`

s/deserialises/would deserialise if canonicity checks were not in place

**Comment by ruhatch** [comment]
Thanks for the reviews @arybczak and @avieth

**Pull request merged**